pkg/image/importer/importer.go

```diff
@@ -0,0 +1,739 @@
+package importer
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	gocontext "golang.org/x/net/context"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/manifest/schema1"
+	"github.com/docker/distribution/registry/api/errcode"
+	"github.com/docker/distribution/registry/api/v2"
+	registryclient "github.com/docker/distribution/registry/client"
+	"github.com/docker/distribution/registry/client/auth"
+	"github.com/docker/distribution/registry/client/transport"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/fielderrors"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/dockerregistry"
+	"github.com/openshift/origin/pkg/image/api"
+	"github.com/openshift/origin/pkg/image/api/dockerpre012"
+)
+
+const (
+	// Add a dockerregistry.Client to the passed context with this key to support v1 Docker registry importing
+	ContextKeyV1RegistryClient = "v1-registry-client"
+)
+
+// Interface loads images into an image stream import request.
+type Interface interface {
+	Import(cxt *AuthContext, isi *api.ImageStreamImport) error
+}
+
+// ImageStreamImport implements an import strategy for Docker images. It keeps a cache of images
+// per distinct auth context to reduce duplicate loads
+type ImageStreamImport struct {
+	maximumTagsPerRepo int
+
+	limiter util.RateLimiter
+
+	imageCache map[*AuthContext]map[location]*api.Image
```

Comment:
Just something I don't know.  Does a map of pointers do what I expect?  Key based on actual memory location, not based on values.